contracts/ForwarderV4.sol

```diff
@@ -0,0 +1,353 @@
+   * @param _from The address of the sender
+   * @param _tokenId The token id of the nft
+   * @param data Additional data with no specified format, sent in call to `_to`
+   */
+  function onERC721Received(
+    address _operator,
+    address _from,
+    uint256 _tokenId,
+    bytes memory data
+  ) external virtual override returns (bytes4) {
+    if (autoFlush721) {
+      IERC721 instance = IERC721(msg.sender);
+      require(
+        instance.supportsInterface(type(IERC721).interfaceId),
+        'The caller does not support the ERC721 interface'
+      );
+      /// this won't work for ERC721 re-entrancy
+      instance.safeTransferFrom(address(this), parentAddress, _tokenId, data);
+    }
+
+    return this.onERC721Received.selector;
+  }
+
+  /**
+   * @notice Method to allow for calls to other contracts. This method can only be called by the allowed address
+   * @param target The target contract address whose method needs to be called
+   * @param value The amount of Ether to be sent
+   * @param data The calldata to be sent
+   */
+  function callFromAllowedAddress(
+    address target,
+    uint256 value,
+    bytes calldata data
+  ) external onlyAllowedAddress returns (bytes memory) {
+    (bool success, bytes memory returnedData) = target.call{ value: value }(
+      data
+    );
+    require(success, 'Allowed address call execution failed');
+
+    return returnedData;
+  }
+
+  /**
+   * @inheritdoc IERC1155Receiver
+   */
+  function onERC1155Received(
+    address _operator,
+    address _from,
+    uint256 id,
+    uint256 value,
+    bytes calldata data
+  ) external virtual override returns (bytes4) {
+    IERC1155 instance = IERC1155(msg.sender);
+    require(
+      instance.supportsInterface(type(IERC1155).interfaceId),
+      'The caller does not support the IERC1155 interface'
+    );
+
+    if (autoFlush1155) {
+      instance.safeTransferFrom(address(this), parentAddress, id, value, data);
+    }
+
+    return this.onERC1155Received.selector;
+  }
+
+  /**
+   * @inheritdoc IERC1155Receiver
+   */
+  function onERC1155BatchReceived(
+    address _operator,
+    address _from,
+    uint256[] calldata ids,
+    uint256[] calldata values,
+    bytes calldata data
+  ) external virtual override returns (bytes4) {
+    IERC1155 instance = IERC1155(msg.sender);
+    require(
+      instance.supportsInterface(type(IERC1155).interfaceId),
+      'The caller does not support the IERC1155 interface'
+    );
+
+    if (autoFlush1155) {
+      instance.safeBatchTransferFrom(
+        address(this),
+        parentAddress,
+        ids,
+        values,
+        data
+      );
+    }
+
+    return this.onERC1155BatchReceived.selector;
+  }
+
+  /**
+   * @inheritdoc IForwarder
+   */
+  function flushTokens(address tokenContractAddress)
+    external
+    virtual
+    override
+    onlyAllowedAddress
+  {
+    ERC20Interface instance = ERC20Interface(tokenContractAddress);
```

Comment:
Is this the best way to check balance of a token (gas efficient)? It still seems a but djanky

Reply:
I think the other option is to use openzeppelin contract IERC20 to figure out the token balance. I tried to call flushTokens using both these ways, the gas utilization seems to be same for both the ways. 

Contract with ERC20Interface: GasLimit: 53857, [Tx](https://goerli.arbiscan.io/tx/0x8010036a28a9cdebe9660a0c1b7cee745c7e0e8b2fd36c2af8ac15d604434de9)
Contract with IERC20: GasLimit: 53857, [Tx](https://goerli.arbiscan.io/tx/0xaa893d52caaae9b14acd4bacc55ce11f9f19153647d86213e99415db98c8505b)

Reply:
if there is no differencre in gaslimit, lets stick to what we have now